lib/nitcorn/proxy.nit

```diff
@@ -0,0 +1,60 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# Copyright 2015 Philippe Pépos Petitclerc <ppeposp@gmail.com>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Provides the `ProxyAction` action, which redirects requests to another 
+# interface
+module proxy
+
+import curl
+import reactor
+
+# A request proxy. All requests received by this action will be redirected to
+# its target. All headers are forwarded including User-Agent.
+#
+# Please note: Requests passed to this action are blocking.
+class ProxyAction
+	super Action
+
+	# Target interface where to dispatch proxied requests
+	var target: Interface
+
+	redef fun answer(http_request, turi) do
+
+		var request = new CurlHTTPRequest("http://{self.target}{turi}")
+		
+		var curl_headers = new HeaderMap
+		
+		for k, v in http_request.header do
+			if k == "User-Agent" then
+				request.user_agent = v
+			else
+				curl_headers[k] = v
+			end
+		end
+
+		var curl_response = request.execute
+		var response = new HttpResponse(200)
+
+		if curl_response isa CurlResponseSuccess then
+			response.body = curl_response.body_str
+		else if curl_response isa CurlResponseFailed then
+			response.body = curl_response.error_msg
```

Comment:
Apache redirects what is received from the remote server. But when there is a CURL error, nothing is received or it is badly formatted. It's probably an unreachable remote server or a config problem. From memory, Apache reports such errors as server errors without much details...

After a look at (Wikipedia)[https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#5xx_Server_Error], there are a few code 5** for proxy servers. HTTP status codes 502, 503, 504 should have their equivalent in CURL and thus could be detected and returned as such. With fallback on 520 when the remote server sent invalid data, and 500 for other errors.

Take a look at the source of the CURL module for the list of errors. Some errors are not yet wrapped, but they only need to be decomented and given a Nity name.